chore(v2): Add initial version of v2 component launcher

neuromage opened the pull request:
**Description of your changes:**


**Checklist:**
- [X] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

neuromage commented:
/assign @Bobgy 
/assign @capri-xiyue 
/cc @chensun 
/cc @james-jwu

numerology commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,438 @@
+		key := fmt.Sprintf(`{{$.inputs.parameters['%s']}}`, k)
+		l.placeholderReplacements[key] = v.ParameterValue
+	}
+
```
Do we want to resolve `{{$}}` placeholder here? Or we can leave a TODO.

neuromage replied:
Added a TODO.

numerology commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,438 @@
+			args[i] = l.placeholderReplacements[v]
+		}
+	}
+
```
Shall we throw an err incase there's any unresolved placeholder left in the args?

capri-xiyue commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,438 @@
+	// Placeholder replacements.
+	pr := make(map[string]string)
+
+	metadata, err := metadata.NewClient(options.MLMDServerAddress, options.MLMDServerPort)
```
Can we rename it to 
```metadataClient, err := metadata.NewClient(options.MLMDServerAddress, options.MLMDServerPort)```?

neuromage replied:
Done.

neuromage commented:
Friendly ping @Bobgy.

Bobgy reviewed (commented):
Great work @neuromage!

I'm leaving some early comments.
Still trying to figure out how to run a sample pipeline using https://github.com/kubeflow/pipelines/pull/5218.

I tried to patch https://github.com/neuromage/pipelines/blob/67ff29013f198cf1bbc0361eb581f6e14273ea0a/sdk/python/kfp/compiler/v2_compatible_compiler_test.py to a pipeline, but it fails on my KFP cluster.

Currently trying to figure out what went wrong

Bobgy commented:
I'm getting this error when running the built sample

```
Failed to successfuly execute component: %vrpc error: code = InvalidArgument desc = Request has no artifact: execution {
  id: 4
  type_id: 10
  last_known_state: COMPLETE
  custom_properties {
    key: "container_image"
    value {
      string_value: "python:3.9"
    }
  }
  custom_properties {
    key: "input:some_int"
    value {
      int_value: 12
    }
  }
  custom_properties {
    key: "input:uri"
    value {
      string_value: "uri-to-import"
    }
  }
  custom_properties {
    key: "kfp_pod_name"
    value {
      string_value: "my-test-pipeline-xqknq-2150358281"
    }
  }
  custom_properties {
    key: "output:output_parameter_one"
    value {
      int_value: 1234
    }
  }
  custom_properties {
    key: "pipeline_name"
    value {
      string_value: "my-test-pipeline"
    }
  }
  custom_properties {
    key: "pipeline_run_id"
    value {
      string_value: "my-test-pipeline-xqknq"
    }
  }
  custom_properties {
    key: "task_name"
    value {
      string_value: "preprocess"
    }
  }
}
artifact_event_pairs {
  event {
    artifact_id: 4
    type: OUTPUT
  }
}
contexts {
  id: 2
  type_id: 8
  name: "my-test-pipeline"
}
contexts {
  id: 5
  type_id: 9
  name: "my-test-pipeline-xqknq"
}
goroutine 1 [running]:
github.com/golang/glog.stacks(0xc001a59200, 0xc001a4e580, 0x4f1, 0x547)
	/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:769 +0xb9
github.com/golang/glog.(*loggingT).output(0x10ff320, 0xc000000003, 0xc001a3e9a0, 0x1082276, 0x7, 0x38, 0x0)
	/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:720 +0x3b3
github.com/golang/glog.(*loggingT).printDepth(0x10ff320, 0x3, 0x1, 0xc0002edf38, 0x2, 0x2)
	/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:646 +0x12d
github.com/golang/glog.(*loggingT).print(...)
	/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:637
github.com/golang/glog.Fatal(...)
	/go/pkg/mod/github.com/golang/glog@v0.0.0-20160126235308-23def4e6c14b/glog.go:1128
main.main()
	/build/cmd/launch/main.go:56 +0x38d
```

Bobgy reviewed (commented):
EDIT: I figured out, the correct command to test is `go test ./...`, the "./" part is necessary.

Another problem:
```
$ go version
go version go1.15.3 darwin/amd64
$ go test '...'
go: github.com/go-kit/kit/examples/stringsvc4 imports
        github.com/nats-io/go-nats: github.com/nats-io/go-nats@v1.8.1: parsing go.mod:
        module declares its path as: github.com/nats-io/nats.go
                but was required as: github.com/nats-io/go-nats
```

Is this problem related to my go version?
After some quick search, I think https://github.com/nats-io/nats.go/issues/478#issuecomment-498849609 fixed the issue for me.

Bobgy commented on v2/launcher_container/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+    GOARCH=amd64
+
+# Copy the code into the container
+COPY cmd /build/cmd
+COPY component /build/component
+COPY metadata /build/metadata
+COPY third_party /build/third_party
+COPY go.mod /build/.
+COPY go.sum /build/.
```
I wonder why we need to copy them one by one.
Can we do `COPY . /build`?

neuromage replied:
Done.

Bobgy commented on v2/metadata/client.go:
```diff
@@ -0,0 +1,365 @@
+func NewClient(serverAddress, serverPort string) (*Client, error) {
+	conn, err := grpc.Dial(fmt.Sprintf("%s:%s", serverAddress, serverPort), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
```
For errors returned from libraries, I believe it's always good practice to wrap it via either:

* https://blog.golang.org/go1.13-errors#TOC_3.3.
* or https://github.com/pkg/errors

Because if we do not wrap the error, there's no call stack trace, so it would be hard to find out where an error occurred when we actually see it.

neuromage replied:
Yep, agreed with you. I did not make this change since the PR was getting big. I'm going to leave a TODO here for this as well.

Bobgy reviewed (commented):
Adding a few more comments as I review, note they are all nitpickings, and do not block merging this PR. We can improve later.

I'll try to grasp the entire structure first and only problems with the structure or interface can potentially block PR merge.

Bobgy commented on v2/component/runtime_info.go:
```diff
@@ -0,0 +1,81 @@
+
+type inputArtifact struct {
+	// Where to read MLMD artifact. File is passed using Argo artifacts.
+	FileInputPath string
```
nit: shall we name it `MetadataInputPath` and `MetadataOutputPath`?

I think that makes it clearer the artifact payload file isn't going into this path.

neuromage replied:
Yes, that makes sense. This change will require us to change compiler too. I'm leaving a TODO here for now, and I'll send a separate PR that renames both. Is that ok with you?

Bobgy commented on v2/component/runtime_info.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+type inputParameter struct {
+	ParameterType  string
```
In argo, we only support string typed parameters. Shall we omit this?

neuromage replied:
That's true, but when storing in MLMD, I've chosen to convert that string back to the intended type (e.g. ints are passed as strings, but stored as ints). WDYT?

neuromage replied:
Yes - you can annotate your function args as `int` or `str` like so:
```
def my_component(arg1: int, arg2: str, ...)
```

Bobgy commented on v2/component/runtime_info.go:
```diff
@@ -0,0 +1,81 @@
+
+	// Generated by launcher.
+	// /tmp/launcher_component_inputs/<name>/data
+	LocalArtifactFilePath string `json:"-"`
```
nit: shall we standardize a term for artifact file? What do you think about calling it payload?

so this becomes `PayloadPath` (I see no reason to repeat artifact,local here, because they are implied by context/the word path.)

neuromage replied:
Ack. Same as above, thinking of doing this in separate PR. Left a TODO.

Bobgy commented on v2/component/runtime_info.go:
```diff
@@ -0,0 +1,81 @@
+}
+
+type outputArtifact struct {
+	ArtifactSchema string
```
discussion: a general naming principle I follow is to use the least numbers of words that can distinguish the name in its context.

what do you think about just `Schema`?

neuromage replied:
SGTM. Left a TODO.

Bobgy commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,440 @@
+	// Placeholder replacements.
+	pr := make(map[string]string)
+
+	metadataClient, err := metadata.NewClient(options.MLMDServerAddress, options.MLMDServerPort)
```
nit: probably do not matter much now, but is the initiated grpc connection going to take up resources in MLMD?

Would it be better to do the following instead?

1. start MLMD connection #1
2. bookkeep execution and artifacts
3. stop connection #1
4. run user command
5. start connection #2
6. ...

neuromage replied:
I think that's an optimization we can make if needed. I don't think these take much resources though. And for quick jobs, it's cheaper to keep the channel around.

Bobgy replied:
I thought the gRPC server has a limit on concurrent connections, if that's not the case, totally fine with it.

Bobgy commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,440 @@
+
+	// Update command.
+	for i, v := range args {
+		if _, ok := l.placeholderReplacements[v]; ok {
```
I think we also need to support args with a placeholder as a substring.
Can you add it as a TODO comment here?

neuromage replied:
Done.

Bobgy reviewed (commented):
Thank you Ajay again! Finished reviewing all the code, this all looks great!

I think all comments are optional. What do you think about letting this merge first and create tracking issues for each of biggerr issues? so that we can start working on the same code base in parallel.

/lgtm

/hold
to let you make the decision

capri-xiyue commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,440 @@
+	pb "github.com/kubeflow/pipelines/v2/third_party/ml_metadata"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"gocloud.dev/blob"
+	_ "gocloud.dev/blob/gcsblob"
```
Yep. The https://gocloud.dev/howto/blob/ supports GCS, S3 and Azure Blob Storage. With this library, we don't need to follow argo to write helper methods for each cloud provider like this https://gocloud.dev/howto/blob/

capri-xiyue replied:
We don't need to write logic to differentiate among GGS, S3, xxx if we just wanna support default authentication/authorization that https://gocloud.dev/howto/blob/  supports.

neuromage commented on v2/launcher_container/Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+# Build the application
+WORKDIR /build
+RUN go build github.com/kubeflow/pipelines/v2/cmd/launch
```
Done.

neuromage commented on v2/third_party/ml_metadata/metadata_source.pb.go:
```diff
@@ -0,0 +1,2607 @@
+
+// Code generated by protoc-gen-go. DO NOT EDIT.
+// versions:
+// 	protoc-gen-go v1.25.0-devel
+// 	protoc        v3.6.1
+// source: metadata_source.proto
```
Thanks, that's helpful. I'll send a follow up PR for updating the Makefile, and also to pin the dependencies of the tools.

neuromage commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,440 @@
+
+	bucket, err := blob.OpenBucket(context.Background(), l.bucketConfig.bucketURL())
+	if err != nil {
+		return fmt.Errorf("Failed to open bucket %q: %v", l.bucketConfig.bucketName, err)
```
Will fix in follow up PR.

neuromage commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,440 @@
+		}
+
+		// copy Artifacts out to remote storage.
+		blobKey, err := l.bucketConfig.keyFromURI(v.URIOutputPath)
```
Done.

neuromage commented on v2/component/launcher.go:
```diff
@@ -0,0 +1,440 @@
+			Uri: &v.URIOutputPath,
+		}
+
+		artifact, err = l.metadataClient.RecordArtifact(ctx, v.ArtifactSchema, artifact)
```
Added a TODO. Agreed. I use this elsewhere as well, may as well combine it.

neuromage commented:
Thanks a lot for the review @Bobgy and @capri-xiyue!

I mostly addressed comments with a bunch of TODOs in the interest of submitting to unblock folks. I'll send follow up PRs to address these over the course of this week.

Bobgy reviewed (commented):
/lgtm
/approve
/unhold

Thank you @neuromage! This is a great start!

Bobgy commented:
/retest
